### PR TITLE
Replace pkg_resources by importlib.resources

### LINKED ISSUE #494: [TODO] Stop importing `pkg_resources`
It's deprecated and causes the following warning:
```console
E   DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
E   Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
```

### PR DESCRIPTION
**Pull request by hynek**
This trades our dependency on setuptools to a conditional one on the importlib-resources backport package.

Fixes #494

### CODE REVIEW

File: src/towncrier/test/helpers.py [hunk 1]
```diff
@@ -20,6 +28,13 @@ def write(path: str | Path, contents: str) -> None:
     p.write_text(contents)
 
 
+def read_pkg_resource(path: str) -> str:
+    """
+    Read *path* from the towncrier package.
+    """
+    return (resources.files("towncrier") / path).read_text()
```
Comment (graingert) [inline src/towncrier/test/helpers.py]:
```suggestion
    return (resources.files("towncrier") / path).read_text("utf8")
```

**Review by graingert — DISMISSED** [review]

File: src/towncrier/_settings/load.py [hunk 2]
```diff
@@ -147,13 +151,14 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
     template = config.get("template", _template_fname)
     if template.startswith("towncrier:"):
         resource_name = "templates/" + template.split("towncrier:", 1)[1] + ".rst"
-        if not pkg_resources.resource_exists("towncrier", resource_name):
+        resource_path = resources.files("towncrier") / resource_name
+        if not resource_path.exists():
```
Comment (graingert) [inline src/towncrier/_settings/load.py]: Traversable objects are not guaranteed to have a `.exists()` and `.resolve()` method, eg if you're running towncrier from a .whl file, I'm not sure why mypy isn't catching this for you

Comment (graingert) [reply src/towncrier/_settings/load.py]: see here for a porting guide for resource_filename: https://importlib-resources.readthedocs.io/en/latest/migration.html#pkg-resources-resource-filename

Comment (hynek) [reply src/towncrier/_settings/load.py]: I was vaguely aware but is that a regression and/or relevant in our use case?

Comment (hynek) [reply src/towncrier/_settings/load.py]: Particularly because we don't consume the file but only pass its path further. as_file makes no sense to me?

Comment (graingert) [reply src/towncrier/_settings/load.py]: yeah the previous resource_filename would extract this template from towncrier zips and wheels and place it in a temporary directory

Comment (hynek) [reply src/towncrier/_settings/load.py]: And then just leave it there??

Comment (graingert) [reply src/towncrier/_settings/load.py]: yeah it would just leave it there, the new api forces a context manager so the file gets cleaned up

Comment (graingert) [reply src/towncrier/_settings/load.py]: mypy didn't catch this because `reveal_type(resource_name) # Any`

File: src/towncrier/_settings/load.py [hunk 3]
```diff
@@ -146,14 +152,20 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
 
     template = config.get("template", _template_fname)
     if template.startswith("towncrier:"):
-        resource_name = "templates/" + template.split("towncrier:", 1)[1] + ".rst"
-        if not pkg_resources.resource_exists("towncrier", resource_name):
+        file_manager = ExitStack()
+        atexit.register(file_manager.close)
```
Comment (graingert) [inline src/towncrier/_settings/load.py]: You're supposed to register this once at module import not every time in some function

File: src/towncrier/_settings/load.py [hunk 4]
```diff
@@ -146,14 +152,20 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
...
+
+        resource_name = f"templates/{template.split('towncrier:', 1)[1]}.rst"
+        resource_path = resources.files("towncrier") / resource_name
+        file_manager.enter_context(resources.as_file(resource_path))
```
Comment (graingert) [inline src/towncrier/_settings/load.py]: you need something like resource_path = file_manager.enter_context(resources.as_file(resource_path))

**Review by graingert — DISMISSED** [review]

**Comment by hynek** [comment]
I think I've addressed everything; CI will be breaking until someone reviews #497 (or fixes it otherwisely).

**Comment by hynek** [comment]
yay green CI

**Review by glyph — APPROVED** [review]

**Pull request merged**